samza-core/src/main/scala/org/apache/samza/checkpoint/OffsetManager.scala

```diff
@@ -261,7 +285,15 @@ class OffsetManager(
         val sspToOffsets = checkpoint.getOffsets
         if(sspToOffsets != null) {
           sspToOffsets.asScala.foreach {
-            case (ssp, cp) => offsetManagerMetrics.checkpointedOffsets.get(ssp).set(cp)
+            case (ssp, cp) => {
+              offsetManagerMetrics.checkpointedOffsets.get(ssp).set(cp)
+
+              // Delete startpoints once a new checkpoint is committed.
+              startpoints.get(taskName) match {
+                case Some(_) => startpointManager.deleteStartpointForTask(ssp, taskName)
```

Comment:
If someone writes a startpoint for an SSP+taskName and then the job model changes, could the store end up with an entry that will never get deleted?

Reply:
TTL would work. If the metadata store itself doesn't give that functionality, you could set a time in the entry and then check if it is "new enough" when handling it.
One undesirable rare case here would be if you don't delete that entry due to a changed job model, and then some time in the future, the mapping happens to work out again such that the job reads the entry and uses a very old startpoint.

Reply:
That's reasonable. Will introduce the notion of staleness and handle it on the reads.